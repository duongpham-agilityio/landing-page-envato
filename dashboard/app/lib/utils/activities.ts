import { AxiosError } from 'axios';

// Store
import { authStore } from '../stores';

// Format
import { formatUppercaseFirstLetter } from '.';

// Interface
import { TActivitiesRequest } from '../interfaces';

// Service
import { MainHttpService, recentActivitiesHttpService } from '../services';

// Constants
import { END_POINTS } from '../constants';

export const handleActivities = async (url: string, actionName: string) => {
  const { user } = authStore();
  try {
    const { data } = await MainHttpService.axiosClient.get(url);

    if (data && user) {
      await recentActivitiesHttpService.post<TActivitiesRequest>(
        END_POINTS.RECENT_ACTIVITIES,
        {
          userId: user.id,
          actionName: actionName,
        },
      );
    }
  } catch (error) {
    const { response } = error as AxiosError<string>;

    throw new Error(formatUppercaseFirstLetter(response?.data));
  }
};
