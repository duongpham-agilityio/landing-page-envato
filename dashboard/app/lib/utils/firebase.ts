import { initializeApp } from 'firebase/app';
import { getAuth } from 'firebase/auth';
import { getStorage } from 'firebase/storage';
import { getFirestore } from 'firebase/firestore';
import { Messaging, getToken } from 'firebase/messaging';

const firebaseConfig = {
  apiKey: process.env.NEXT_PUBLIC_FIREBASE_API_KEY,
  authDomain: process.env.NEXT_PUBLIC_FIREBASE_AUTH_DOMAIN,
  projectId: process.env.NEXT_PUBLIC_FIREBASE_PROJECT_ID,
  storageBucket: process.env.NEXT_PUBLIC_FIREBASE_STORAGE_BUCKET,
  messagingSenderId: process.env.NEXT_PUBLIC_FIREBASE_SENDER_ID,
  appId: process.env.NEXT_PUBLIC_FIREBASE_APP_ID,
  measurementId: process.env.NEXT_PUBLIC_FIREBASE_MEASUREMENT_ID,
};

initializeApp(firebaseConfig);

const VAPID_KEY = process.env.NEXT_PUBLIC_FIREBASE_VAPID;

export const requestForToken = (messaging: Messaging) =>
  getToken(messaging, { vapidKey: VAPID_KEY })
    .then((currentToken) => {
      if (currentToken) {
        console.log('current token for client: ', currentToken);

        return currentToken;
        // Perform any other neccessary action with the token
      } else {
        // Show permission request UI
        console.log(
          'No registration token available. Request permission to generate one.',
        );
      }
    })
    .catch((err) => {
      console.log('An error occurred while retrieving token. ', err);
    });

// Initialize Firebase
export const app = initializeApp(firebaseConfig);
export const auth = getAuth();
export const storage = getStorage();
export const db = getFirestore();
