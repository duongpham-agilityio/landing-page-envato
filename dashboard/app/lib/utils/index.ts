export * from './transaction';
export * from './time';
export * from './user';
export * from './url';
export * from './helpers';
export * from './text';
export * from './localStorage';
export * from './image';
export * from './prefetch';
export * from './toast';
