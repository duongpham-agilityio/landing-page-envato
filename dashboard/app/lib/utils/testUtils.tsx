import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { ReactNode } from 'react';

const queryClient = new QueryClient();

export const renderQueryProviderTest = (component: JSX.Element) => {
  const { render } = testLibReactUtils;
  return render(
    <QueryClientProvider client={queryClient}>{component}</QueryClientProvider>,
  );
};

export const queryProviderWrapper = ({ children }: { children: ReactNode }) => (
  <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>
);
