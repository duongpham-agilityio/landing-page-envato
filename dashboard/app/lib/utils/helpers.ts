// Constants
import { REGEX, DOTS, ERROR_MESSAGES } from '@/lib/constants';

// Types
import { FormatPaginationParams, PaginationTableType } from '@/lib/interfaces';

export const formatNumberButton = (numberOfPage: number): string[] =>
  Array.from({ length: numberOfPage }).map((_, index: number): string =>
    (index + 1).toString(),
  );

export const formatPageArray = ({
  totalPage = 0,
  currentPage = 0,
}: PaginationTableType): string[] => {
  // If the number of pages is less than or equal to 5, display all page buttons
  if (totalPage <= 5) {
    return formatNumberButton(totalPage);
  }

  // If the current page is near the start (pages 1 to 3),
  // Display the first 3 pages and then DOTS and the last page
  if (currentPage <= 3) {
    return [...formatNumberButton(3), DOTS, totalPage.toString()];
  }

  // If the current page is near the end (within the last 3 pages),
  // display the first page, DOTS, and the last 3 pages
  if (currentPage >= totalPage - 2) {
    return ['1', DOTS, ...formatNumberButton(totalPage).splice(-3)];
  }

  // If the current page is in the middle,
  // Display the first page, DOTS, currentPage, DOTS, and the last page
  return ['1', DOTS, currentPage.toString(), DOTS, totalPage.toString()];
};

export const formatPagination = ({
  totalCount,
  limit,
  currentPage,
}: FormatPaginationParams): string[] => {
  const totalPage = Math.ceil(totalCount / limit);

  return formatPageArray({ totalPage, currentPage });
};

export const validatePassword = (value: string) => {
  if (!value) {
    return ERROR_MESSAGES.FIELD_REQUIRED('Password');
  }

  if (!REGEX.LENGTH_IS_EIGHT.test(value)) {
    return ERROR_MESSAGES.PASS_WORD_SHORT;
  }

  if (!REGEX.PASSWORD.test(value)) {
    return ERROR_MESSAGES.PASS_WORD_WEAK;
  }

  return true;
};

export const formatUppercaseFirstLetter = (value = ''): string =>
  value.charAt(0).toUpperCase() + value.slice(1);

export const formatDecimalInput = (value = ''): string => {
  const validData = /(^[0-9])[^.]*((?:\.\d*)?)/.exec(
    value.replace(/[^(\d|.)]/g, ''),
  );

  const formatValue = validData ? validData[0] : '';

  return formatValue;
};

export const formatAmountNumber = (value: string): string => {
  if (!value) {
    return value;
  }

  if (Number.isNaN(parseFloat(value))) {
    return '';
  }

  const dotIndex = value.indexOf('.');
  const decimalValue = dotIndex > -1 ? value.slice(dotIndex) : '';
  const newValue =
    dotIndex > -1
      ? value.slice(0, dotIndex).replaceAll(',', '')
      : value.replaceAll(',', '');

  const newValueFormat = newValue.replace(/\B(?=(\d{3})+(?!\d))/g, ',');

  if (
    !REGEX.AMOUNT_PATTERN.test(newValue) ||
    !REGEX.DECIMAL_PATTERN.test(decimalValue)
  ) {
    return `${newValueFormat}.${decimalValue
      .substring(0, 3)
      .replaceAll(/[.]/g, '')}`;
  }

  return decimalValue ? `${newValueFormat}${decimalValue}` : newValueFormat;
};

/**
 * Remove amount format ex: 12,345.00 -> 12345
 * @param amount string
 * @returns number
 */
export const removeAmountFormat = (amount: string): number =>
  +amount.replaceAll(',', '');

export const parseFormattedNumber = (value: string): number => {
  if (!value) {
    return 0;
  }

  const cleanValue = String(value).replace(REGEX.FORMAT_NUMBER, '');

  const parsedNumber = parseFloat(cleanValue);

  if (Number.isNaN(parsedNumber)) {
    return 0;
  }

  return parsedNumber;
};

/**
 * Format number rg: 12345 -> 12,345.00 if isOmitDecimals = false or 12,345 if isOmitDecimals = true
 * @param number
 * @param isOmitDecimals
 * @returns Number after format
 */
export const formatDecimalNumber = (
  number = 0,
  isOmitDecimals: boolean = false,
): string => {
  const formattedNumber = isOmitDecimals
    ? Math.round(number).toString()
    : number.toFixed(2);

  const numberWithCommas = formattedNumber.replace(
    /\B(?=(\d{3})+(?!\d))/g,
    ',',
  );

  return numberWithCommas;
};

// Format for typing numbers in input
export const formatAllowOnlyNumbers = (input: string): string => {
  if (!input) {
    return ''; // or handle the case when input is undefined or null
  }

  // Replace any non-digit character with an empty string
  return input.replace(/[^0-9]/g, '');
};

export const getStatusColor = (status: string): string => {
  switch (status.toLowerCase()) {
    case 'online':
      return 'green.500';
    case 'offline':
      return 'gray.500';
    default:
      return 'gray.500';
  }
};

export const isWindowDefined = () => typeof window !== 'undefined';
