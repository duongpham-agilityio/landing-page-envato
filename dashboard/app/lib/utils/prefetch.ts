// Libs
import { QueryClient } from '@tanstack/react-query';

// Services
import { MainHttpService } from '@/lib/services';

export const prefetchStatistical = async <T>(
  endPoint: string,
  queryClient: QueryClient,
) => {
  await queryClient.prefetchQuery({
    queryKey: [endPoint],
    queryFn: async () =>
      (await MainHttpService.get<T>({ path: endPoint })).data,
  });
};
