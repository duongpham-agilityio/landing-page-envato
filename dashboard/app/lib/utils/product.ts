import dayjs from 'dayjs';

// Types
import { Status, TProduct } from '@/lib/interfaces';

// Utils
import { formatDecimalNumber, formatUppercaseFirstLetter } from '.';

// Constants
import { IMAGES, TIME_FORMAT } from '../constants';

/**
 * Convert data show for home page
 * @param products
 * @returns
 */
export const formatProductResponse = (products: TProduct[] = []) =>
  products.map((product) => {
    const {
      _id,
      name,
      amount,
      currency,
      createdAt,
      description,
      imageURLs,
      stock,
    } = product;

    return {
      id: _id,
      name: formatUppercaseFirstLetter(name),
      description: formatUppercaseFirstLetter(description),
      date: dayjs(createdAt).format(TIME_FORMAT),
      imageURLs: imageURLs || IMAGES.BIG_AVATAR.url,
      amount: `${currency}${formatDecimalNumber(+amount)}`,
      stock: `${stock}`,
      status: stock > 0 ? Status.IN_STOCK : Status.SOLD,
    };
  });
