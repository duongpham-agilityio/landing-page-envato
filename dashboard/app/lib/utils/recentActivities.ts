import dayjs from 'dayjs';

// Types
import { TRecentActivities } from '@/lib/interfaces';

// Utils
import { formatUppercaseFirstLetter } from '.';

// Constants
import { TIME_FORMAT } from '../constants';

/**
 * Convert data show for home page
 * @param recentActivities
 * @returns
 */
export const formatRecentActivitiesResponse = (
  recentActivities: TRecentActivities[] = [],
) =>
  recentActivities.map((activity) => {
    const { _id, actionName, email, createdAt } = activity;

    return {
      _id: _id,
      name: formatUppercaseFirstLetter(actionName),
      email,
      date: dayjs(createdAt).format(TIME_FORMAT),
    };
  });
