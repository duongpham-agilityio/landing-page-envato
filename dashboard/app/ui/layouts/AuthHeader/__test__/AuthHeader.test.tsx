import '@testing-library/jest-dom';

// component
import { AuthHeader } from '../..';

// Utils
import { renderQueryProviderTest } from '@/lib/utils/testUtils';

describe('Auth Header component render', () => {
  const renderComponent = (path?: string) =>
    renderQueryProviderTest(<AuthHeader title="test" pathName={path} />);

  it('Should render match with snapshot.', () => {
    const { container } = renderComponent();
    expect(container).toMatchSnapshot();
  });

  it('Should be displayed if pathName is different from the path', () => {
    const { container } = renderComponent('/some/other/path');
    expect(container).toBe('/some/other/path');
  });
});
