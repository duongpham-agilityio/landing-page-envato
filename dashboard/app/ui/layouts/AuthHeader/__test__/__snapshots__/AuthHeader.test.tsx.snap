// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Auth Header component render Should render match with snapshot. 1`] = `
<div>
  <header
    class="chakra-stack css-tl3ftk"
  >
    <h1
      class="chakra-heading css-3ub5ab"
    >
      test
    </h1>
    <p
      class="chakra-text css-1cl46s8"
    >
      Send, spend and save smarter
    </p>
  </header>
  <span
    hidden=""
    id="__chakra_env"
  />
</div>
`;
